### PR TITLE
[FLINK-30998] Add optional exception handler to flink-connector-opensearch

### PR DESCRIPTION
**Pull request by lilyevsky**
See comments on https://github.com/apache/flink-connector-opensearch/pull/8 .
These are the same changes re-targeted to main branch.

### CODE REVIEW

**Comment by lilyevsky** [comment]
@reta The validator did not like the PR title. Not sure what it wants. Could you please fix it?

**Comment by reta** [comment]
> @reta The validator did not like the PR title. Not sure what it wants. Could you please fix it?

@lilyevsky I would love to but I don't have enough permissions to do that. Could you modify your title please to be: `
[FLINK-30998] Add optional exception handler to flink-connector-opensearch`?

File: flink-connector-opensearch/pom.xml [hunk 1]
```diff
@@ -36,7 +36,7 @@ under the License.
 
 	<!-- Allow users to pass custom connector versions -->
 	<properties>
-		<opensearch.version>1.3.0</opensearch.version>
+		<opensearch.version>2.5.0</opensearch.version>
```
Comment (reta) [inline flink-connector-opensearch/pom.xml:39]: Please revert, Opensearch 2.x needs JDK-11

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSink.java [hunk 2]
```diff
@@ -59,19 +59,22 @@
     private final BulkProcessorConfig buildBulkProcessorConfig;
     private final NetworkClientConfig networkClientConfig;
     private final DeliveryGuarantee deliveryGuarantee;
+    private FailureHandler failureHandler;
```
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSink.java:62]: `final` please

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/FailureHandler.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.opensearch.sink;
+
+import java.io.Serializable;
+
+/** Handler to process failures. */
+public interface FailureHandler extends Serializable {
```
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/FailureHandler.java:24]: Could you please add @FunctionalInterface to this interface (for usability perspective).

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java [hunk 4]
```diff
@@ -285,6 +286,18 @@ public OpensearchSinkBuilder<IN> setAllowInsecure(boolean allowInsecure) {
...
 
+    /**
+     * Allows to set custom failure handler. If not set, or set to null, then the default behavior
+     * is preserved: throwing a runtime exception upon receiving a failure.
+     *
+     * @param failureHandler
+     * @return
+     */
+    public OpensearchSinkBuilder<IN> setFailureHandler(FailureHandler failureHandler) {
+        this.failureHandler = failureHandler;
```
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java:297]: Please `checkNotNull(failureHandler);`

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java [hunk 5]
```diff
@@ -343,7 +346,11 @@ private void extractFailures(BulkRequest request, BulkResponse response) {
         if (chainedFailures == null) {
             return;
         }
-        throw new FlinkRuntimeException(chainedFailures);
+        if (failureHandler == null) {
```
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:349]: Dealing with`null` is better to be avoided: we could use the default `failureHandler`:

```java
private static final FailureHandler DEFAULT_FAILURE_HANDLER = ex -> throw new FlinkRuntimeException(ex);
```

and never deal with nullability / `if`s

File: flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java [hunk 6]
```diff
@@ -266,7 +266,8 @@ private OpensearchWriter<Tuple2<Integer, String>> createWriter(
                         null,
                         true),
                 metricGroup,
-                new TestMailbox());
+                new TestMailbox(),
+                null);
```
Comment (reta) [inline flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:270]: `DEFAULT_FAILURE_HANDLER` would help here to eliminate `null`

File: flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java [hunk 7]
```diff
@@ -207,7 +205,7 @@ public void testItemFailureRethrownOnCheckpointAfterFlush() throws Throwable {
                                 1,
                                 OpType.INDEX,
                                 new IndexResponse(
-                                        new ShardId("test", "-", 0), "_doc", "1", 0, 0, 1, true))));
+                                        new ShardId("test", "-", 0), "1", 0, 0, 1, true))));
```
Comment (reta) [inline flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:208]: Please revert

**Comment by lilyevsky** [comment]
@reta It looks like it was wrong idea for me to make changes on ....2.5.0 branch and then create a PR targeted to main.
Let me do it all over again, it would be simpler. I will make changes in my repo in main branch and create a PR to main.

**Comment by lilyevsky** [comment]
I will close this PR and create another one. I did make the fixes according to your suggestions.

**Pull request closed without merging**